### ISSUE TITLE
`reduce_by_key` fails with custom reduction

### ISSUE THREAD

**Issue by upsj** [ROCmSoftwarePlatform/rocThrust#352]
I'm using reduce_by_key to implement a segmented argmin. The following piece of code fails with a access violationt with ROCm 5.6.0 (since little changed since then, I assume it's still failing), the equivalent code works fine with NVIDIA thrust. I haven't been able to debug the corresponding issues, since I don't have access to a system with sufficient driver support for `rocgdb`. My educated guess is that the reduction is being evaluated for invalid/out-of-bounds argument values, which leads to out-of-bounds accesses to `degree`

```cpp
#include <thrust/device_vector.h>
#include <thrust/reduce.h>

template <typename IndexType>
struct node_min_degree_reduction {
    __device__ __forceinline__ IndexType operator()(IndexType u, IndexType v)
    {
        return thrust::make_pair(degree[u], u) < thrust::make_pair(degree[v], v)
                   ? u
                   : v;
    }

    const IndexType* degree;
};

int main() {
        using IndexType = int;
        thrust::device_vector<int> k(1000, 0);
        thrust::device_vector<int> v{1000, 1};
        thrust::device_vector<int> res1(1);
        thrust::device_vector<int> res2(1);
        thrust::device_vector<int> degrees(1000, 1);
        thrust::reduce_by_key(k.begin(), k.end(), v.begin(), res1.begin(), res2.begin(), thrust::equal_to<IndexType>{},
                          node_min_degree_reduction<IndexType>{degrees.data().get()});

}
```

**Comment by mfep** [comment]
Hi @upsj, and thank you for the apt bug report! I can confirm that this reproducer fails on ROCm version 5.6.

It is very likely that this bug was fixed in https://github.com/ROCm/rocPRIM/pull/434 (rocPRIM is the algorithmic library behind rocThrust on the HIP backend), which is shipped as part of the ROCm 6.0 release - the provided reproducer passes on that version of rocThrust/rocPRIM.

Can you please confirm that your use case works as expected on ROCm 6.0?

**Comment by upsj** [comment]
That fixed it, thanks
